### PR TITLE
enable wandb sweep integration

### PR DESCRIPTION
**Pull request by ayl**
Allows for integration with the wandb sweep hyperparameter search feature. 

Wandb currently does not support nested parameters thus the arguments list needs to be flattened and rebuilt.

### CODE REVIEW

**Comment by ayl** [comment]
Here is an example sweep file that works with the tutorial code:

```
program: deepy.py
method: bayes
metric:
  name: validation/lm_loss
  goal: minimize
parameters:
  optimizer.params.lr:
    min: 0.0001
    max: 0.1
command:
  - ${env}
  - ${interpreter}
  - ${program}
  - "train.py"
  - "configs/small.yml"
  - "configs/local_setup.yml"
```

To launch the sweep you would run:
`wandb sweep sweep.yaml`

which will give you a command that you use launch each of the agents.

**Review by StellaAthena — APPROVED** [review]
Excellent work! Thank you.

Before this is merged, can you write up some documentation and add it to the readme?

**Comment by ayl** [comment]
@StellaAthena Done! I added the sweep.yml example to the README and some documentation around how to use it.

File: README.md [hunk 1]
```diff
@@ -312,6 +312,41 @@ In addition to storing logs locally, we provide built-in support for two popular
 
 EleutherAI is currently using [Weights & Biases to record our experiments](https://wandb.ai/eleutherai/neox). If you are logged into Weights & Biases on your machine&mdash;you can do this by executing `wandb login`&mdash;your runs will automatically be recorded. There are two optional fields associated with Weights & Biases: <code><var>wandb_group</var></code> allows you to name the run group and <code><var>wandb_team</var></code> allows you to assign your runs to an organization or team account.
 
+## wandb sweeps 
+
+Hyperparameter searches can also be automated using the [wandb sweeps](https://docs.wandb.ai/guides/sweeps). This allows for automated seaches of hyperparameters. Nested configuration values from `configs/` need to be separated by periods (eg. `optimizer.params.lr`) and underscores must be used instead of dashes.
```
Comment (Mistobaan) [inline README.md:317]: *searches

Comment (ayl) [reply README.md]: Thanks for catching that typo!

**Comment by StellaAthena** [comment]
I was able to [create a sweep successfully](https://wandb.ai/eleutherai/gpt-neox/sweeps/9a0sippz/overview?workspace=user-stellaathena) but when I try to execute it I am getting the following error
```Traceback (most recent call last):
  File "train.py", line 26, in <module>
    neox_args.configure_distributed_args()
AttributeError: 'NoneType' object has no attribute 'configure_distributed_args'```

I've made the linked sweep open, so anyone can try it out / edit it.

**Comment by StellaAthena** [comment]
> @StellaAthena Could you post your
> 
> * configs/small.yml
> * configs/local_setup.yml
> 
> so I can replicate your setup? When I tried it, I think it was complaining that the `save_interval` was missing from the configs.
> 
> With respect to your earlier question, the wandb agent runs after deepy.py is called so it can control most/all of the parameters after deepspeed is started (within train.py). I am not sure I have the hardware to replicate your setup and your specific question.

Ohhh I didn’t notice that you were pointing at `local_setup`… let me try it again with `eleutherai_cluster`

**Pull request closed without merging**